File: iOS Applications/Objective-C Fundamentals/GSM/GSM/Display.m
//
//  Display.m
//  GSM

#import "Display.h"

@implementation Display {
    CGFloat _size;
    NSInteger _numberOfColors;
}

-(instancetype)initWithSize:(CGFloat)size
          andNumberOfColors:(NSInteger)numberOfColors {
    self = [super init];
    
    if (self) {
        self.size = size;
        self.numberOfColors = numberOfColors;
    }
    
    return self;
}

+(Display *)displayWithSize:(CGFloat)size
           andNuberOfColors:(NSInteger)numberOfColors {
    return [[Display alloc] initWithSize: size
                       andNumberOfColors: numberOfColors];
}

-(CGFloat)size {
    return _size;
}

-(void)setSize:(CGFloat)size {
    if (size <= 0) {
        NSException *ex = [[NSException alloc] initWithName: @"Size out of range" reason: @"Size must be bigger than 0" userInfo: nil];
        [ex raise];
    }

    _size = size;
}

-(NSInteger)numberOfColors {
    return _numberOfColors;
}

- (void)setNumberOfColors:(NSInteger)numberOfColors {
    if (numberOfColors <= 0) {
        NSException *ex = [[NSException alloc] initWithName: @"Number of colors out of range" reason: @"Number of colors must be bigger than 0" userInfo: nil];
        [ex raise];
    }
    
    _numberOfColors = numberOfColors;
}

- (NSString *)description {
    return [NSString stringWithFormat: @"Size: %.1f, Number of colors: %ld", self.size, (long)self.numberOfColors ];
}

@end

File: C# OOP/Defining-Classes-Part-One/GSM/Display.cs
﻿namespace GSM
{
    using System;

    public class Display
    {
        private double size;
        private int numberOfColors;

        public Display(double size, int numberOfColors)
        {
            this.Size = size;
            this.NumberOfColors = numberOfColors;
        }

        public double Size
        {
            get
            {
                return this.size;
            }
            set
            {
                if (value <= 0)
                {
                    throw new ArgumentOutOfRangeException("Size of display must be positive number");
                }

                this.size = value;
            }
        }

        public int NumberOfColors
        {
            get
            {
                return this.numberOfColors;
            }
            set
            {
                if (value <= 0)
                {
                    throw new ArgumentOutOfRangeException("Number of colors must be bigger than 0");
                }

                this.numberOfColors = value;
            }
        }

        public override string ToString()
        {
            return String.Format("Size: {0}, Number of colors: {1}", this.Size, this.NumberOfColors); 
        }

    }
}
